## Clean and Analyze eBay Used Car Listings

We have a dataset of 50,000 data points from eBay Kleinanzeigen, using a secion of the German eBay website.

* Go through the used car listings dataset and:
    * Explore the data
    * Identify any patterns
    * Remove non-digit characters
    * Convert columns
    * Rename column (if required)

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos.info()
autos.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset has 20 columns, most of which are stored as strings and some stored as integers. 
These columns contain null values:
    * vehicleType
    * gearbox
    * model
    * fuelType
    * notRepairedDamage
* Dates are stored as strings. 
* Name column has underscores and forward slash (/)
* Price - change price to int and rename the column to include the currency
* Odometer - same as price, except rename column to include km

In [3]:
# Clean the columns!!!!
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

* Convert the column names from camelcase to snakecase
* Add a better description to some columns

In [4]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test",
                "vehicle_type", "registration_year", "gearbox", "power_ps", "model",
                "odometer", "registration_month", "fuel_type", "brand",
                "unrepaired_damage", "ad_created", "num_photos", "postal_code",
                "last_seen"]
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Explore the data and see what other cleaning tasks need to be done.
* Any columns that mostly have one value may be dropped
* Numeric data stored as text needs to be cleaned

In [5]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Observations:
* num_photos columns needs further investigation
* seller and offer_type columns have nearly all the same values

In [6]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

It appears that all the values in the num_photos column are 0. Need to drop this column, as well as, seller and offer_type.

In [7]:
#autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [8]:
# Remove non-numeric characters from price 
# Convert the column to a numeric dtype
# autos["price"] = (autos["price"]
#                     .str.replace("$","")
#                     .str.replace(",","")
#                     .astype(int)
#                 )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [9]:
# Remove non-numeric characters from odometer
# Rename column to odometer_km
# autos["odometer"] = (autos["odometer"]
#                         .str.replace("km","")
#                         .str.replace(",","")
#                         .astype(int)
#                     )
# autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Continue the exploration of data and specifically looking at data that just doesn't look right. Looking at "odometer_km" and "price" columns.

In [10]:
# Look at the price column first
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(15)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

# Price observations
* There are 1421 car listings with the price of \$0 
* The maximum price is \$100,000,000 which seems way too high
* The prices are all whole numbers (rounded)

In [11]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

It seems that the last reasonable (heh) price is \$350,000 and jumps up to \$999,990, which is a 65% increase.

In [13]:
autos["price"].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are quite a few number of listings that are under \$30, including the 1421 listings we mentioned earlier at \$0. However, I do not think we need to be alarmed at the amount of \$1 listings as the seller may have started the bid at \$1. 

Will be removing the listings that are at \$0 and the listings above \$350,000 as they are too unrealistic. 

In [14]:
autos = autos[autos["price"].between(1, 350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Need to convert "date_crawled", "last_seen", and "ad_created" to be numerical data type

In [16]:
autos[["date_crawled", "ad_created", "last_seen" ]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

First 10 characters of each column is the date YYYY-MM-DD
Need to extract these date values

In [18]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Daily crawl

In [19]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [20]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [21]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum value is 1000, when cars did not exist yet and the maximum value is 9999, an unforeseeable year.

In [22]:
autos["registration_year"].value_counts().sort_index(ascending=False).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64

Need to get rid of the "Years" above 2019

In [23]:
autos["registration_year"].value_counts().sort_index().head(15)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
Name: registration_year, dtype: int64

Need to get rid of the years below 1910

In [26]:
autos = autos[autos["registration_year"].between(1900,2019)]
# autos["registration_year"].describe()
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.065012
2005    0.060480
1999    0.059677
2004    0.055680
2003    0.055598
2006    0.055001
2001    0.054300
2002    0.051210
1998    0.048676
2007    0.046905
Name: registration_year, dtype: float64

Most of the cars have been registered for the past ~20 years

In [28]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212813
opel              0.108641
bmw               0.108641
mercedes_benz     0.095787
audi              0.085858
ford              0.069667
renault           0.047894
peugeot           0.029457
fiat              0.025996
seat              0.018931
skoda             0.016068
nissan            0.015264
mazda             0.015223
smart             0.014296
citroen           0.014111
toyota            0.012586
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009043
mini              0.008611
mitsubishi        0.008178
honda             0.007993
kia               0.007107
alfa_romeo        0.006612
porsche           0.005912
suzuki            0.005891
chevrolet         0.005665
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001566
jaguar            0.001524
trabant           0.001380
r

The top five brands are all German manufacturers. Lada has the smallest percentage of the listings. We will focus on the brands that represent more than 5% of the total listing.

In [29]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [32]:
brand_mean_prices = {}

for brand in common_brands:
    brand_name = autos[autos["brand"]== brand]
    mean_price = brand_name["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'audi': 9212,
 'bmw': 8261,
 'ford': 3728,
 'mercedes_benz': 8526,
 'opel': 2941,
 'volkswagen': 5333}

* Audi, BMW, and Mercedes-Benz are the most expensive cars
* Ford and Opel are the least expensive cars
* Volkswagen is in the middle

In [33]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
audi                 9212
bmw                  8261
ford                 3728
mercedes_benz        8526
opel                 2941
volkswagen           5333

In [34]:
# Exploring Mileage
brand_mean_mileage = {}

for brand in common_brands:
    brand_name = autos[autos["brand"] == brand]
    mean_mileage = brand_name["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [36]:
brand_info = pd.DataFrame(mean_mileage, columns=["mean_mileage"])
brand_info

mean_mileage
bmw                  132682
mercedes_benz        130848
audi                 129492
opel                 129452
volkswagen           128955
ford                 124349

In [37]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132682        8261
mercedes_benz        130848        8526
audi                 129492        9212
opel                 129452        2941
volkswagen           128955        5333
ford                 124349        3728

The car mileages do not vary that much, just a small trend where the more expensive cars have a higher mileage than the less expensive cars. 

In [52]:
# autos.head(10)

In [47]:
# Convert the vehicle_type column from German translation to English
mapping_dict = {
    "bus": "van",
    "limousine": "sedan",
    "kombi": "sedan",
    "kleinwagen": "hatchback",
    "cabrio": "convertible",
    "suv": "suv",
    "coupe": "coupe"
}

autos["vehicle_type"] = autos["vehicle_type"].map(mapping_dict)
# autos.head(10)

In [49]:
# Convert the gearbox column from German translation to English
gear_map = {
    "manuell": "manual",
    "automatik": "automatic"
}

autos["gearbox"] = autos["gearbox"].map(gear_map)
# autos.head(10)

In [50]:
# Convert the unrepaired_damage column from German translation to English

ud_map = {
    "nein": "no",
    "ja": "yes"
}

autos["unrepaired_damage"] = autos["unrepaired_damage"].map(ud_map)
# autos.head(10)

In [53]:
# Convert the fuel_type column from German translation to English
fuel_map = {
    "diesel": "diesel",
    "benzin": "gas",
    "lpg": "lpg"
}

autos["fuel_type"] = autos["fuel_type"].map(fuel_map)
autos.head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offer_type  price  ab_test vehicle_type  registration_year gearbox  \
0  privat    Angebot   5000  control          NaN               2004     NaN   
1  privat    Angebot   8500  control          NaN               1997     NaN   
2  privat    Angebot   8990     test          NaN               2009     NaN   
3  privat    Angebot   4350  control          NaN               2007     NaN   
4  privat    Angebot   1350     test          NaN               2003     NaN   
5  privat    Angebot   7900     test          NaN               2006     NaN   
6  privat    Angebot    300     test          NaN               1995     NaN   
7  privat    Angebot   1990  control          NaN               1998     NaN   
8  privat    Angebot    250     test          NaN               2000     NaN   
9  privat    Angebot    590  control          NaN               1997     NaN   

   power_ps    model  odometer_km  registration_month fuel_type       brand  \
0       158   andere       150000                   3       lpg     peugeot   
1       286      7er       150000                   6       NaN         bmw   
2       102     golf        70000                   7       NaN  volkswagen   
3        71   fortwo        70000                   6       NaN       smart   
4         0    focus       150000                   7       NaN        ford   
5       150  voyager       150000                   4    diesel    chrysler   
6        90     golf       150000                   8       NaN  volkswagen   
7        90     golf       150000                  12    diesel  volkswagen   
8         0    arosa       150000                  10       NaN        seat   
9        90   megane       150000                   7       NaN     renault   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0               NaN  2016-03-26 00:00:00           0        79588   
1               NaN  2016-04-04 00:00:00           0        71034   
2               NaN  2016-03-26 00:00:00           0        35394   
3               NaN  2016-03-12 00:00:00           0        33729   
4               NaN  2016-04-01 00:00:00           0        39218   
5               NaN  2016-03-21 00:00:00           0        22962   
6               NaN  2016-03-20 00:00:00           0        31535   
7               NaN  2016-03-16 00:00:00           0        53474   
8               NaN  2016-03-22 00:00:00           0         7426   
9               NaN  2016-03-16 00:00:00           0        15749   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50  
5  2016-04-06 09:45:21  
6  2016-03-23 02:48:59  
7  2016-04-07 03:17:32  
8  2016-03-26 18:18:10  
9  2016-04-06 10:46:35

In [54]:
# Convert the dates to be uniform numeric data so "2016-03-21" becomes the integer 20160321
autos["date_crawled"] = (autos["date_crawled"]
                         .str[:10]
                         .str.replace("-","")
                         .astype(int)
                        )
autos["date_crawled"].head(10)
    

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
5    20160321
6    20160320
7    20160316
8    20160322
9    20160316
Name: date_crawled, dtype: int64

In [55]:
autos["ad_created"] = (autos["ad_created"]
                          .str[:10]
                          .str.replace("-","")
                          .astype(int)
                      )
autos["ad_created"].head(10)

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
5    20160321
6    20160320
7    20160316
8    20160322
9    20160316
Name: ad_created, dtype: int64

In [56]:
autos["last_seen"] = (autos["last_seen"]
                         .str[:10]
                         .str.replace("-","")
                         .astype(int)
                     )
autos["last_seen"].head(10)

0    20160406
1    20160406
2    20160406
3    20160315
4    20160401
5    20160406
6    20160323
7    20160407
8    20160326
9    20160406
Name: last_seen, dtype: int64

In [58]:
# autos["name"].head(20)

In [65]:
# Find the most common brand/model combinations
pd.crosstab(autos["brand"], autos["model"], dropna=False)

model          100  145  147  156  159  1_reihe  1er  200  2_reihe  300c  \
brand                                                                      
alfa_romeo       0    5   85   89   33        0    0    0        0     0   
audi            58    0    0    0    0        0    0    1        0     0   
bmw              0    0    0    0    0        0  534    0        0     0   
chevrolet        0    0    0    0    0        0    0    0        0     0   
chrysler         0    0    0    0    0        0    0    0        0    11   
citroen          0    0    0    0    0        0    0    0        0     0   
dacia            0    0    0    0    0        0    0    0        0     0   
daewoo           0    0    0    0    0        0    0    0        0     0   
daihatsu         0    0    0    0    0        0    0    0        0     0   
fiat             0    0    0    0    0        0    0    0        0     0   
ford             0    0    0    0    0        0    0    0        0     0   
honda            0    0    0    0    0        0    0    0        0     0   
hyundai          0    0    0    0    0        0    0    0        0     0   
jaguar           0    0    0    0    0        0    0    0        0     0   
jeep             0    0    0    0    0        0    0    0        0     0   
kia              0    0    0    0    0        0    0    0        0     0   
lada             0    0    0    0    0        0    0    0        0     0   
lancia           0    0    0    0    0        0    0    0        0     0   
land_rover       0    0    0    0    0        0    0    0        0     0   
mazda            0    0    0    0    0       10    0    0        0     0   
mercedes_benz    0    0    0    0    0        0    0    0        0     0   
mini             0    0    0    0    0        0    0    0        0     0   
mitsubishi       0    0    0    0    0        0    0    0        0     0   
nissan           0    0    0    0    0        0    0    0        0     0   
opel             0    0    0    0    0        0    0    0        0     0   
peugeot          0    0    0    0    0      156    0    0      610     0   
porsche          0    0    0    0    0        0    0    0        0     0   
renault          0    0    0    0    0        0    0    0        0     0   
rover            0    0    0    0    0        0    0    0        0     0   
saab             0    0    0    0    0        0    0    0        0     0   
seat             0    0    0    0    0        0    0    0        0     0   
skoda            0    0    0    0    0        0    0    0        0     0   
smart            0    0    0    0    0        0    0    0        0     0   
subaru           0    0    0    0    0        0    0    0        0     0   
suzuki           0    0    0    0    0        0    0    0        0     0   
toyota           0    0    0    0    0        0    0    0        0     0   
trabant          0    0    0    0    0        0    0    0        0     0   
volkswagen       0    0    0    0    0        0    0    0        0     0   
volvo            0    0    0    0    0        0    0    0        0     0   

model           ...    wrangler  x_reihe  x_trail  x_type  xc_reihe  yaris  \
brand           ...                                                          
alfa_romeo      ...           0        0        0       0         0      0   
audi            ...           0        0        0       0         0      0   
bmw             ...           0      302        0       0         0      0   
chevrolet       ...           0        0        0       0         0      0   
chrysler        ...           0        0        0       0         0      0   
citroen         ...           0        0        0       0         0      0   
dacia           ...           0        0        0       0         0      0   
daewoo          ...           0        0        0       0         0      0   
daihatsu        ...           0        0        0       0         0      0   
fiat            ...      